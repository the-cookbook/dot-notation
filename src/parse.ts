import pick from './pick';
import parseKey from './parse-key';
import getArrayIndex from './utils/get-array-index';
import is from './utils/is';
import merge from './utils/merge';
import createPathBreadcrumb from './utils/create-path-breadcrumb';

const compile = (
  source: Record<string, unknown> | unknown[],
  instructions: string[],
  value: unknown,
): Record<string, unknown> | unknown[] => {
  const data = is.array(source) ? [...source] : { ...source };

  if (is.object(data)) {
    return merge(data, parseKey(instructions.join('.'), value) as Partial<Record<string, unknown>>);
  }

  const { 1: idx, index } = getArrayIndex(instructions.join('.'));

  if (+idx > data.length) {
    throw new RangeError(
      `Expected array index for path "${instructions.join('.')}" to be "${data.length}" but found instead "${idx}"`,
    );
  }

  if (is.nullOrUndefined(data[+idx])) {
    data.push(...(parseKey(instructions.splice(index).join('.'), value) as unknown[]));
  } else {
    const hasChild = instructions.length > 1;

    const result = hasChild ? compile((data as unknown[])[+idx] as unknown[], instructions.slice(1), value) : value;

    if (is.object(result)) {
      data[+idx] = { ...result };
    } else if (is.array(result)) {
      data[+idx] = [...result];
    } else {
      data[+idx] = result;
    }
  }

  return data;
};

/**
 * Parse object from dot notation
 * @template T
 * @param {object} source - Dot notation object
 * @returns {object}
 * @param source
 * @return {T|T[]}
 */
const parse = <T>(source: Record<string, unknown>): T extends [] ? T[] : T => {
  const paths = Object.keys(source);

  let result: unknown = getArrayIndex(createPathBreadcrumb(paths[0])[0]) ? [] : {};

  for (let i = 0; i < paths.length; i += 1) {
    const path = paths[i];
    const hasArrayNotation = getArrayIndex(path);
    const value = source[path];

    let parsedValue = parseKey(path, value);

    if (hasArrayNotation) {
      const commonPath = path.substr(0, hasArrayNotation.index);
      const workingPath = createPathBreadcrumb(path.replace(commonPath, ''));
      const workingNode: unknown[] = commonPath ? pick(result, commonPath) || [] : (result as unknown[]);

      parsedValue = compile(workingNode, workingPath, value);

      if (commonPath) {
        parsedValue = parseKey(commonPath, parsedValue);
      }
    }

    if (is.array(parsedValue)) {
      result = parsedValue;
    }

    result = merge<Record<string, unknown>>(
      result as Partial<Record<string, unknown>>,
      parsedValue as Partial<Record<string, unknown>>,
    );
  }

  return result as T extends [] ? T[] : T;
};

export default parse;
