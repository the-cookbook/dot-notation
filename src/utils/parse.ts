const toKebabCase = (value: string, limiter = '-'): string => {
  if (!value) {
    return '';
  }

  return value
    .match(/[A-Z]{2,}(?=[A-Z][a-z]+[0-9]*|\b)|[A-Z]?[a-z]+[0-9]*|[A-Z]|[0-9]+/g)
    .map((x) => x.toLowerCase())
    .join(limiter);
};

const toCamelCase = (value: string, lowerCamelCase = true): string => {
  if (!value) {
    return '';
  }

  return value
    .match(/[A-Z]{2,}(?=[A-Z][a-z]+[0-9]*|\b)|[A-Z]?[a-z]+[0-9]*|[A-Z]|[0-9]+/g)
    .map((x, idx) => {
      if (lowerCamelCase && idx === 0) {
        return x.toLowerCase();
      }

      return x.substr(0, 1).toUpperCase() + x.substr(1);
    })
    .join('');
};

const objectKeysCase = <T extends Record<string, unknown> = Record<string, unknown>>(
  source: T,
  caseModifier: (value: string) => string = toCamelCase,
): T => {
  return Object.keys(source).reduce((list, key) => ({ ...list, [caseModifier(key)]: source[key] }), {} as T);
};

export default {
  toKebabCase,
  toCamelCase,
  objectKeysCase,
};
