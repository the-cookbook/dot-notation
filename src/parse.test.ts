import parse from './parse';

describe('parse()', () => {
  it('should match snapshot', () => {
    expect(parse).toMatchSnapshot();
  });

  describe('exceptions', () => {
    it('should throw an error when array index is inconsistent', () => {
      const source = {
        'person.name.firstName': 'John',
        'person.name.lastName': 'Doe',
        'person.address[0].street': 'Infinite Loop',
        'person.address[1].city': 'Cupertino',
        'person.address[4].postalCode': 95014,
      };

      expect(() => parse(source)).toThrow(RangeError);
    });

    it('should parse overwrite array path', () => {
      const source = {
        '[0]': 'loren',
        '[1][0]': 'hello',
        '[1][1]': 'world',
        '[1]': 'ipsum', // expects to overwrite the array created above
        '[2]': 'amet',
        '[3]': 'safir',
        '[4]': 'asinha',
      };

      const expectation = ['loren', 'ipsum', 'amet', 'safir', 'asinha'];

      expect(parse(source)).toStrictEqual(expectation);
    });
  });

  describe('object path', () => {
    it('should parse simple object path accordingly', () => {
      const source = {
        'person.name': 'John',
        'person.lastName': 'Doe',
      };

      expect(parse(source)).toStrictEqual({
        person: {
          name: 'John',
          lastName: 'Doe',
        },
      });
    });

    it('should parse complex object path accordingly', () => {
      const source = {
        'person.name.firstName': 'John',
        'person.name.lastName': 'Doe',
        'person.address.street': 'Infinite Loop',
        'person.address.city': 'Cupertino',
        'person.address.postalCode': 95014,
      };

      expect(parse(source)).toStrictEqual({
        person: {
          name: {
            firstName: 'John',
            lastName: 'Doe',
          },
          address: {
            street: 'Infinite Loop',
            city: 'Cupertino',
            postalCode: 95014,
          },
        },
      });
    });
  });

  describe('array path', () => {
    it('should parse array path accordingly', () => {
      const source = {
        '[0]': 'Bacon',
        '[1][0]': 'ipsum',
        '[1][1][0]': 'dolor',
        '[1][1][1][0]': 'amet',
      };

      const expectation = ['Bacon', ['ipsum', ['dolor', ['amet']]]];

      expect(parse(source)).toStrictEqual(expectation);
    });
  });

  describe('nested path', () => {
    it('should parse object path accordingly', () => {
      const source = {
        'person.addresses[0].street': 'Infinite Loop',
        'person.addresses[].city': 'Cupertino',
        'person.addresses[0].postalCode': 95014,
        'person.name.firstName': 'John',
        'person.name.lastName': 'Doe',
      };

      expect(parse(source)).toStrictEqual({
        person: {
          name: {
            firstName: 'John',
            lastName: 'Doe',
          },
          addresses: [
            {
              street: 'Infinite Loop',
              city: 'Cupertino',
              postalCode: 95014,
            },
          ],
        },
      });
    });

    it('should parse object with array indexes', () => {
      const source = {
        'person.name.firstName': 'John',
        'person.name.lastName': 'Doe',
        'person.addresses[0].street': 'Infinite Loop',
        'person.addresses[0].city': 'Cupertino',
        'person.addresses[0].postalCode': 95014,
        'person.addresses[1].street': '1600 Amphitheatre',
        'person.addresses[1].city': 'Mountain View',
        'person.addresses[1].postalCode': 94043,
      };

      expect(parse(source)).toStrictEqual({
        person: {
          name: {
            firstName: 'John',
            lastName: 'Doe',
          },
          addresses: [
            {
              street: 'Infinite Loop',
              city: 'Cupertino',
              postalCode: 95014,
            },
            {
              street: '1600 Amphitheatre',
              city: 'Mountain View',
              postalCode: 94043,
            },
          ],
        },
      });
    });

    it('should parse object with deep arrays', () => {
      const source = {
        'person.name.firstName': 'John',
        'person.name.lastName': 'Doe',
        'person.addresses[0].street': 'Infinite Loop',
        'person.addresses[0].city': 'Cupertino',
        'person.addresses[0].postalCode': 95014,
        'person.addresses[1][0].street': '1600 Amphitheatre',
        'person.addresses[1][1].city': 'Mountain View',
        'person.addresses[1][2].postalCode': 94043,
      };

      expect(parse(source)).toStrictEqual({
        person: {
          name: {
            firstName: 'John',
            lastName: 'Doe',
          },
          addresses: [
            {
              street: 'Infinite Loop',
              city: 'Cupertino',
              postalCode: 95014,
            },
            [
              {
                street: '1600 Amphitheatre',
              },
              {
                city: 'Mountain View',
              },
              {
                postalCode: 94043,
              },
            ],
          ],
        },
      });
    });
  });
});
