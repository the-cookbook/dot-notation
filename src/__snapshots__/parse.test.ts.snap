// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`parse() should match snapshot 1`] = `[Function]`;
